hbase-client/src/main/java/org/apache/hadoop/hbase/client/CatalogReplicaLoadBalanceReplicaSimpleSelector.java

```diff
@@ -0,0 +1,273 @@
+  private final long STALE_CACHE_TIMEOUT_IN_MILLISECONDS = 3000; // 3 seconds
+  private final int STALE_CACHE_CLEAN_CHORE_INTERVAL = 1500; // 1.5 seconds
+
+  /**
+   * StaleLocationCacheEntry is the entry when a stale location is reported by an client.
+   */
+  private static final class StaleLocationCacheEntry {
+    // replica id where the stale location comes from.
+    private int fromReplicaId;
+
+    // timestamp in milliseconds
+    private long timestamp;
+
+    private byte[] endKey;
+
+    StaleLocationCacheEntry(final int metaReplicaId, final byte[] endKey) {
+      this.fromReplicaId = metaReplicaId;
+      this.endKey = endKey;
+      timestamp = System.currentTimeMillis();
+    }
+
+    public byte[] getEndKey() {
+      return this.endKey;
+    }
+
+    public int getFromReplicaId() {
+      return this.fromReplicaId;
+    }
+    public long getTimestamp() {
+      return this.timestamp;
+    }
+
+    @Override
+    public String toString() {
+      return new ToStringBuilder(this)
+        .append("endKey", endKey)
+        .append("fromReplicaId", fromReplicaId)
+        .append("timestamp", timestamp)
+        .toString();
+    }
+  }
+
+  private static final class StaleTableCache {
+    private final ConcurrentNavigableMap<byte[], StaleLocationCacheEntry> cache =
+      new ConcurrentSkipListMap<>(BYTES_COMPARATOR);
+  }
+
+  private final ConcurrentMap<TableName, StaleTableCache> staleCache;
+  private int numOfReplicas;
+  private final AsyncConnectionImpl conn;
+  private TableName tableName;
+
+  CatalogReplicaLoadBalanceReplicaSimpleSelector(TableName tableName, AsyncConnectionImpl conn) {
+    staleCache = new ConcurrentHashMap<>();
+    this.conn = conn;
+    this.tableName = tableName;
+
+    // This numOfReplicas is going to be lazy initialized.
+    this.numOfReplicas = -1;
+    // Start connection's chore service in case.
+    this.conn.startChoreService();
+    this.conn.getChoreService().scheduleChore(getCacheCleanupChore(this));
+  }
+
+  /**
+   * When a client runs into RegionNotServingException, it will call this method to
+   * update Selector's internal state.
+   * @param loc the location which causes exception.
+   * @param fromReplicaId the replica id where the stale location comes from.
+   */
+  public void notifyOnError(HRegionLocation loc, int fromReplicaId) {
+    StaleTableCache tableCache =
+      computeIfAbsent(staleCache, loc.getRegion().getTable(), StaleTableCache::new);
+    byte[] startKey = loc.getRegion().getStartKey();
+    tableCache.cache.putIfAbsent(startKey,
+      new StaleLocationCacheEntry(fromReplicaId, loc.getRegion().getEndKey()));
+    LOG.debug("Add entry to stale cache for table {} with startKey {}, {}",
+      loc.getRegion().getTable(), startKey, loc.getRegion().getEndKey());
+  }
+
+  private int getRandomNonPrimaryReplicaId() {
+    if (numOfReplicas < 0) {
+      try {
+        Admin admin = conn.toConnection().getAdmin();
```

Comment:
I think here we'd better make use of the Locator related API, or use ConnectionRegistry directly? Usually it is not a good idea to involve admin related API in the critical read/write path.

Reply:
And do we need to refresh this value? For a long running client, the HBase cluster could change the meta replica count.

Reply:
Let me update. Since there is already a chore service running. I can add a chore which checks the updated meta replica count every 1 min.